### PR TITLE
#166495052 Implement Role-based Authorization

### PR DESCRIPTION
**Pull request by henryade**
#### What does this PR do?
Authorize users based on specified roles
#### Description of Task to be completed?
Implement Role-based Authorization to authorize users based on the roles specified
#### How should this be manually tested?
it will be tested when it's used in a route that has the verify token middleware
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166495052](https://www.pivotaltracker.com/story/show/166495052)
#### Screenshots (if appropriate)
Route:
image

Admin-User:
image
image

User with no role:
image
image

#### Questions:
N/A

### CODE REVIEW

File: server/middlewares/authorizeUser.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import models from '../database/models';
+import responseMessage from '../helpers';
+
+const { UserRole, Role } = models;
+
+export default async (request, response, next, roleArray) => {
+  try {
+    const { id } = request.userData;
```
Comment (Lundii) [inline server/middlewares/authorizeUser.js]: What is userData?

File: server/middlewares/authorizeUser.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    const userrole = await UserRole.findOne({
+      where: { userId: id },
+      include: [Role]
+    });
+    if (!userrole) responseMessage(response, 403, { message: 'UnAuthorized User. Role not found' });
+    const { roleName } = userrole.Role;
+    if (roleArray.indexOf(roleName) === -1) responseMessage(response, 403, { message: 'UnAuthorized User' });
```
Comment (Lundii) [inline server/middlewares/authorizeUser.js]: What I understand is happening here is that roleName is checked if it exits in the roleArray. What happens when a user with a "user role" tried to do something meant for the "adminRole" Will the person be authorized?

File: server/middlewares/authorizeUser.js (hunk 1 above)
Comment (henryade) [reply server/middlewares/authorizeUser.js]: Userdata is attached to the request from in the verify token middleware. So I am getting the id from there.

File: server/middlewares/authorizeUser.js (hunk 2 above)
Comment (henryade) [reply server/middlewares/authorizeUser.js]: The array specifies who can access the route

File: server/middlewares/authorizeUser.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+const { UserRole, Role } = models;
+
+export default async (request, response, next, roleArray) => {
+  try {
+    const { id } = request.userData;
+    const userrole = await UserRole.findOne({
+      where: { userId: id },
+      include: [Role]
+    });
+    if (!userrole) responseMessage(response, 403, { message: 'UnAuthorized User. Role not found' });
```
Comment (Lundii) [inline server/middlewares/authorizeUser.js]: I think the convention is to make the message lower case

File: server/middlewares/authorizeUser.js (hunk 2 above)
Comment (Lundii) [inline server/middlewares/authorizeUser.js]: The message here too

Comment (henryade) [reply server/middlewares/authorizeUser.js]: done

File: server/middlewares/authorizeUser.js (hunk 3 above)
Comment (henryade) [reply server/middlewares/authorizeUser.js]: done

**Review by Lundii — COMMENTED** [review]
Looks good to me

**Review by Lundii — DISMISSED** [review]

**Comment by allebd** [comment]
How can it be tested, using it in a route as stated did not work, when I tested it or can you direct me on how to test it

**Review by KvNGCzA — APPROVED** [review]

**Review by allebd — APPROVED** [review]

**Comment by Adekoreday** [comment]
Nice work
you need to remove the comments in the admin routes
you need to write test for roles that exist and role that does not 
because the test coverage dropped by 4%

**Review by Adekoreday — CHANGES_REQUESTED** [review]
Nice work
remove the comments in admin.js
you need to write more test to for role that exist and role that does not exist
the test coverage dropped by 4%

**Comment by henryade** [comment]
> Nice work
> remove the comments in admin.js
> you need to write more test to for role that exist and role that does not exist
> the test coverage dropped by 4%

there's no route to put the middleware yet. the comment in the admin.js page is for when there is a route. you will know how to use it.

**Review by Adekoreday — APPROVED** [review]

**Pull request merged**